File: 디자인패턴/template method/Animal.java
package TemplateMethodPattern;

public abstract class Animal {

    // 템플릿 메서드
    public void playWithOwner(){
        System.out.println("헤에에이 컴온~");
        play();
        runSomething();
        System.out.println("귯결");
    }

    // 추상메서드 (필수 구현)
    abstract void play();

    // Hook메서드 (선택 구현)
    void runSomething(){
        System.out.println("꼬리 살랑살랑");
    }
}

File: 디자인패턴/template method/Animal.kt
abstract class Animal {
    // template method
    fun playWithOwner() {
        println("귀염둥이 이리온~")
        play()
        runSomething()
        println("잘했어")
    }

    // abstract method
    abstract fun play()

    // hook method
    open fun runSomething() {
        println("꼬리 살랑살랑~")
    }
}
